src/mono/mono/mini/mini-amd64.c

```diff
@@ -1014,17 +1015,23 @@ get_call_info (MonoMemPool *mp, MonoMethodSignature *sig)
 		if (mono_method_signature_has_ext_callconv (sig, MONO_EXT_CALLCONV_SWIFTCALL)) {
 			MonoClass *swift_self = mono_class_try_get_swift_self_class ();
 			MonoClass *swift_error = mono_class_try_get_swift_error_class ();
+			MonoClass *swift_indirect_result = mono_class_try_get_swift_indirect_result_class ();
 			MonoClass *swift_error_ptr = mono_class_create_ptr (m_class_get_this_arg (swift_error));
 			MonoClass *klass = mono_class_from_mono_type_internal (sig->params [i]);
-			if (klass == swift_self && sig->pinvoke) {
+			if (klass == swift_indirect_result)
+				cinfo->swift_indirect_result_index = i;
+
+			if ((klass == swift_self || klass == swift_indirect_result) && sig->pinvoke) {
 				guint32 size = mini_type_stack_size_full (m_class_get_byval_arg (klass), NULL, sig->pinvoke && !sig->marshalling_disabled);
 				g_assert (size == 8);
 
 				ainfo->storage = ArgValuetypeInReg;
 				ainfo->pair_storage [0] = ArgInIReg;
 				ainfo->pair_storage [1] = ArgNone;
 				ainfo->nregs = 1;
-				ainfo->pair_regs [0] = GINT32_TO_UINT8 (AMD64_R13);
+				// The indirect result is RAX on AMD64. However, since RAX is used as a scratch register, 
```

Comment:
If RAX is clobbered inside interp trampolines, I think we should simply avoid clobbering it there. This looks quite confusing.

Reply:
I think it's not only interp trampolines that use RAX, a lot of other functions inside `mini-amd64.c` use RAX as well.

Reply:
Any idea of a specific use of RAX that prevents us from allocating swift indirect result directly into RAX ?

Reply:
For example varargs calls (check `amd64_handle_varargs_call`) seem to have the address of the call in RAX and move it to R11 for the call. However, I'm not aware of specific use of RAX, I think it is mostly used as a temp/scratch register. There is over 50 occurrences of hard-coded `AMD64_RAX` inside `mini-amd64.c` so it's possible there are some other use-cases.

Reply:
I think RAX is implicitly assigned in `mini-codegen.c` within the `alloc_reg` function, as it is the first one. If this is true, we should have the callinfo information at this point, and we can try to to mask RAX and use the next available register.

Reply:
Anyway, this issue is not a new one, the same pattern is already implemented in https://github.com/dotnet/runtime/blob/3e15b2898386637034d738f6e63204f1e721cb45/src/mono/mono/mini/mini-amd64.c#L4610-L4614

I believe it is important to fix this, but perhaps it should be addressed in a follow-up and targeted for .NET 10. I think this would be quite risky for the Preview 7 release.

Reply:
I'm fine with not doing this as part of this PR. It is indeed high risk, if it involves the reg allocator. At the end of the day, amd64 is probably not the highest priority arch, so we might end up with code as is. I would update the comment in the code so it is clear why we are doing this. My understanding is that this is not because we just randomly use RAX as temporary scratch in the code, but rather that in m2n wrappers, the reg allocator is free to  assign the ftn_ptr to it, conflicting with our required use of it as an argument reg.